### DISCUSSION TITLE
Artisan Make Trait

### DISCUSSION THREAD

**Discussion by hamzaelmaghari** [laravel/framework#50825] [Q&A]
### Hello devs,

Laravel 11 come with new `make` command let's talk first about `php artisan make:trait Example`, by default the generated trait will be created in `App` directory, instead of `App\Traits`.

My question is don't you find this weird!?

![tenor](https://github.com/laravel/framework/assets/84163115/40a82063-bca9-46c8-bbb3-d348a174849b)

**Comment by ahinkle** [comment]
No, these are intended to be flexible in design as they aren't framework-specific. https://github.com/laravel/framework/pull/50492

You can customize the directory on creation. e.g. `make:trait Models/Traits/Example`

**Reply by hamzaelmaghari** [reply]
As a Laravel developer I used to see requests in `Requests` folder and controllers in `Controllers` folder etc... as Laravel is known for its clean setup and directories.

**Reply by ghost** [reply]
You know it is faster to just copy paste an existing class and rename it. I never generate files using commands , except migrations because of their timestamp and name and then copy paste from an old migration the structure. Main reason is the lack of PSR12 as Laravel follows only PSR2.

**Reply by hamzaelmaghari** [reply]
I do the same in my phpstorm but in a different way, I import the class first then I create it by inlay hints on hover 🫡 as this is not an issue, it is just ecosystem quality checking.

**Reply by vitorbizarra** [reply]
Wouldn't it be interesting if, by default, traits were created within the Traits directory if a directory for it is not specified? For example:

```
artisan make:trait Example // will be created in App\Traits
artisan make:trait Models/Traits/Example // will be created in App\Models\Traits
```

I've been thinking about this and would like to know what people with more experience in the framework think. Would that be a problem? I can't think of any case where it would be necessary to create the trait in the root of the app directory.

**Reply by ahinkle** [reply]
@vitorbizarra It already does that if the `App/Traits` folder exists. If the folder doesn't exist, it will create the trait under the default namespace in `App`.

**Reply by hamzaelmaghari** [reply]
> @vitorbizarra It already does that if the `App/Traits` folder exists. If the folder doesn't exist, it will create the trait under the default namespace in `App`.

Nice, the first time I knew that, though it doesn't match Laravel philosophy.

This algorithm should be reversed, the command should create the directory `Traits` if not existing.

**Reply by ahinkle** [reply]
As mentioned above, traits aren't specific to the framework unlike other generators. It makes sense not to make it the default.

**Reply by syntafin** [reply]
> @vitorbizarra It already does that if the `App/Traits` folder exists. If the folder doesn't exist, it will create the trait under the default namespace in `App`.

I can say, it doesnt do this. ``App\Traits`` already exists, but Traits made via Artisan still created in ``App`` :(

**Reply by ahinkle** [reply]
@syntafin Make sure you're on the latest version of the framework. works for me

**Reply by syntafin** [reply]
> @syntafin Make sure you're on the latest version of the framework. works for me

it is, but some strange things, tested it via shell (outside IDE) and works as "expected".... need to check if IDE use the artisan file inside the project root :)
